common/state/trie.cpp

```diff
@@ -469,3 +468,199 @@ void pstate::trie_node::operate_trie_root(
     //         in the case of failure, any modification is discarded, the transaction in progress will not succeed,
     //         no new state is generated, and the request processing can (and will) start over from the last state
 }
+
+namespace pdo
+{
+namespace state
+{
+    class recursive_item
+    {
+    public:
+        trie_node node;
+        bool go_next; // true: proceed with next offset; false: proceed with child offset
+    };
+}
+}
+
+void pstate::trie_node::operate_trie_non_recursive(
+    data_node_io& dn_io, const kv_operation_e operation, const ByteArray& kvkey, const ByteArray& in_value, ByteArray& out_value)
+{
+    std::vector<recursive_item> trie_recursion_stack;
+    unsigned int depth = 0;
+    unsigned int loopn = 0;
+
+    // the first entry of the first data node is the trie root
+    // if the trie contains data then the root has a next node
+    // if the trie is empty then the next node is null/empty
+    recursive_item ri;
+    recursive_item ri_next;
+    ri.node.location.block_offset_ = {dn_io.block_warehouse_.get_root_block_num(), data_node::data_begin_index()};
+    trie_node::read_trie_node(dn_io, ri.node.location.block_offset_, ri.node);
+    //initialize the recursion stack with the root node
+    ri.go_next = true;
+    trie_recursion_stack.push_back(ri);
+    //append next uninitialized node
+    ri_next.node.location.block_offset_ = ri.node.node.next_offset;
+    trie_recursion_stack.push_back(ri_next);
+
+    while(1)
+    {
+        // reference to last item in stack
+        // notice: the referenced node may not be initialized, it must be created if location is null,or read otherwise
+        recursive_item& ri = trie_recursion_stack.back();
+
```

Comment:
changed to list
(the references were/are used correctly, since they are not required once the structure is update; I agree however that is easy to mess up)